core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -2267,7 +2267,7 @@ private void trimLabels(@CheckForNull Set<LabelAtom> includedLabels) {
         this.getNodes().forEach(n -> nodeLabels.addAll(n.getAssignedLabels()));
         for (Iterator<Label> itr = labels.values().iterator(); itr.hasNext();) {
             Label l = itr.next();
-            if (includedLabels == null || includedLabels.contains(l)) {
+            if (includedLabels == null || includedLabels.contains(l) || l.matches(includedLabels)) {
```

Comment:
I am missing something here.
`Jenkins.getAssigneLabeles` returns a `Set<LabelAtom>`   
`getNodes()` returns a `List<Node>` and `Node.getAssignedLabeles()` also returns a `Set<LabelAtom>`

So whilst the Set of `nodeLabels` is declared as `Set<Label>` all the contents should be `LabelAtoms`

A LabelAtom should by definition be a single atomic label  so there should be no need for `l.matches(Colleciton)` here as a LabelAtom should only ever match the same Label Atom..

also `includedLabels.contains()` should be using equality for the set and `equals` is defined in Label to use `matches()` which works on the name of the label which is a string which in the case of LabelAtom is the label.

so if `n.getAssignedLabels` is not returning atoms of LabelAtoms then the code calling this on Node removal `void trimLabels(Node... nodes)` .

This would seem to point at something being wrong somewhere else?

Reply:
doesn't it make sense to revert https://github.com/jenkinsci/jenkins/pull/5882 for now then? most likely the problem is originated there

Reply:
alas not - as that fixed a major regression in the previous change here.

Reply:
maybe this is because the equals method in Label uses getClass for comparison?

Reply:
OK so `Label.equals(Object other)` (and `LabelAtom.equals` which defaults to the parent as it is not over ridden)
check class equality

https://github.com/jenkinsci/jenkins/blob/6edca13517f484e595f6f7341198239bc01ce8d5/core/src/main/java/hudson/model/Label.java#L514-L521

This will fail if anyone subclasses `LabelAtom`

Docker Swarm does 
https://github.com/jenkinsci/docker-swarm-plugin/blob/d6e0dc1fd4717dc178cf2ab81c8e7a5820fbf9fd/src/main/java/org/jenkinsci/plugins/docker/swarm/DockerSwarmLabelAssignmentAction.java#L49-L62

as does a few others.

Now `LabelAtoms` are supposed to be the actual Atom, and there may well be other cases where the code has assumed that LabelAtom is never subclassed when it is not.

Reply:
unless overridden in subclasses (which it is not as far as I can see) this means a Label subtype will only be equal to any other Label subtype if it's name matches, irrespective of any other things.

so whilst this code is correct there appears to be other potential bugs lurking here (and I still fail to see how for the ec2 plugin that is not subclassing LabelAtom that this fixes the issue.

Reply:
I am a maintainer of anka-build plugin and we are seeing similar issues (as explained in the issue related). We are also not overriding Label or LabelAtom, so this is not a specific EC2 thing (nor does it seem to have anything to do with overwriting these classes)

> includedLabels.contains() should be using equality for the set and equals is defined in Label to use matches() ...

I just ran some tests locally, the includedLabels.contains() does NOT fire the overridden equals method in Label

Reply:
> I just ran some tests locally, the includedLabels.contains() does NOT fire the overridden equals method in Label

contains is on a `HashSet` so the contains should be using the `hascode()` of the `LabelAtom` which is defined in `Label` as `name.hascode()`

Could you provide the example of the label that matches but is not in contains so I can understand this better?

Reply:
Furthermore, there can be cases, like the one we are discussing, where the equals method in Label would return false and not fire the underlying match method if comparing LabelAtom and LabelExpression.
Now, I am not sure if this behavior is expected or not, but wouldn't using 'instanceof Label' more appropriate in this equals implementation?

Reply:
> Could you provide the example of the label that matches but is not in contains so I can understand this better, as something seems wrong and I am trying to get to the root cause.

I'm sorry, I don't understand what you asking of me exactly.
You can see the exact label names and pipeline code I am using in the issue that this PR is trying to solve

Reply:
> > Could you provide the example of the label that matches but is not in contains so I can understand this better, as something seems wrong and I am trying to get to the root cause.
> 
> I'm sorry, I don't understand what you asking of me exactly. You can see the exact label names and pipeline code I am using in the issue that this PR is trying to solve

I do not have an ec2 cloud to test with - I have another cloud implementations and they appear to be working as expected.

when you said:

> I just ran some tests locally, the includedLabels.contains() does NOT fire the overridden equals method in Label

I possibly incorrectly assumed you had stepped though this code in a debugger with the cloud that is not behaving as you expected.  I was just asking you to provide the details of the `LabelAtom` that did not match in the `contains` but did match in the `matches`, along with the contents of the `includedLabels.

I will try and make time to setup the ec2 plugin and debug this to work out what is going on.

Reply:
Ahh, ok, is this what you mean?

I have one job running asking for 2 agents from label "lbl1||lbl2" in parallel
Four agents are being spun up, all of them lbl1

includedLabels: [cloud-lbl1-MPXKN, lbl1]
nodeLabels: [cloud-lbl1-gY3Hz, built-in, cloud-lbl1-MPXKN, cloud-lbl1-si1oa, lbl1, cloud-lbl1-7JsJ3]
l: lbl1||lbl2
contains returns false while matches returns true

Reply:
ok, so walking this through with @Vlatombe the queue is not considering a task buildable as the labelExpression has no clouds (or nodes).  
This should fix that, but there is also another issue here in that label expressions are never garbage collected so long as something that can satisfy them exists, even if they where only used by a job/build that has long gone.  This is an existing issue and I will file an ticket for that.

At the same time it would seem that this is not limited to clouds but should also be able to be reproduced with a job and static agents (if the job had label expression `foo||bar`, and was run when Jenkins had no Nodes, then a user created a static node with the label `foo` - it would not build immediatly that node was added but would wait for the full maintainance task to kick in (`includedLabels == null`).

This should make a unit test possible without having to write a mock cloud implementation.

Reply:
https://issues.jenkins.io/browse/JENKINS-67642 FTR